File: android/app/src/main/java/app/stepsapp/domain/Aggregate.kt
package app.stepsapp.domain

import java.time.LocalDate
import java.time.YearMonth

/** 集計の単位。 */
enum class Period { DAY, WEEK, MONTH, YEAR }

/**
 * 集計結果。
 *
 * @param key          期間の識別子。週は「その週の月曜日」、月は YYYY-MM、年は YYYY
 * @param average      記録がある日だけを分母にした平均（記録日平均）
 * @param total        合計
 * @param daysRecorded 記録がある日数。0歩と記録された日も含む
 * @param daysInPeriod その期間の暦日数。欠損量を見せるために使う
 */
data class Bucket(
    val key: String,
    val average: Double,
    val total: Long,
    val daysRecorded: Int,
    val daysInPeriod: Int,
)

/**
 * その日が属する週の月曜日。
 *
 * サーバ側 `server/src/aggregate.ts` の SQL と**同じ規則**でなければならない
 * （`date(d, '-' || ((strftime('%w', d) + 6) % 7) || ' days')`）。
 * 年をまたぐ週を分断しないため、週は「月曜日の日付」で識別する。
 */
fun weekStart(date: LocalDate): LocalDate =
    date.minusDays(((date.dayOfWeek.value + 6) % 7).toLong())

/**
 * 日次の記録を期間ごとにまとめる。
 *
 * **平均の分母は記録がある日だけ。** 未計測の日を0歩として平均を下げない。
 * 0歩と記録された日は分母に含める。
 *
 * @param stepsByDate 日付 -> 歩数（記録がある日だけ）
 * @return 新しい期間が先頭に来る順
 */
fun aggregate(stepsByDate: Map<String, Long>, period: Period): List<Bucket> {
    if (stepsByDate.isEmpty()) return emptyList()

    val grouped = stepsByDate.entries.groupBy { (dateText, _) ->
        when (period) {
            Period.DAY -> dateText
            // 週だけは月曜日を求めるのに暦が要る。ほかは文字列の切り出しで足りる。
            // 全日を LocalDate.parse していたころは、期間を切り替えるたびに
            // 1,200 回の暦計算が走って画面が引っかかっていた
            Period.WEEK -> weekStart(LocalDate.parse(dateText)).toString()
            Period.MONTH -> dateText.substring(0, 7)
            Period.YEAR -> dateText.substring(0, 4)
        }
    }

    return grouped.map { (key, entries) ->
        val total = entries.sumOf { it.value }
        Bucket(
            key = key,
            average = total.toDouble() / entries.size,
            total = total,
            daysRecorded = entries.size,
            daysInPeriod = daysInPeriod(key, period),
        )
    }.sortedByDescending { it.key }
}

fun daysInPeriod(key: String, period: Period): Int = when (period) {
    Period.DAY -> 1
    Period.WEEK -> 7
    Period.MONTH -> YearMonth.parse(key).lengthOfMonth()
    Period.YEAR -> if (LocalDate.of(key.toInt(), 1, 1).isLeapYear) 366 else 365
}

/**
 * 直前の期間との比較。
 *
 * **期間の途中では「同じ経過日数までの前期間」と比べる。** 月初に
 * 「先月より大幅に少ない」と出るのは不公平で、意味のある比較にならないため。
 *
 * @param current  今の期間の記録（日付 -> 歩数）
 * @param previous 前の期間の記録
 * @param elapsedDays 今の期間の経過日数。前期間もこの日数ぶんだけ切り出して比べる
 * @return 平均の差（今 - 前）。どちらかに記録が無ければ null
 */
fun compareAverages(
    current: Map<String, Long>,
    previous: Map<String, Long>,
    elapsedDays: Int,
): Double? {
    if (current.isEmpty()) return null

    val previousClipped = previous.entries
        .sortedBy { it.key }
        .take(elapsedDays)
    if (previousClipped.isEmpty()) return null

    val currentAvg = current.values.average()
    val previousAvg = previousClipped.sumOf { it.value }.toDouble() / previousClipped.size
    return currentAvg - previousAvg
}

File: web/aggregate.js
// 集計ロジック。ブラウザでそのまま動くよう依存を持たない。
//
// **週の境界の規則は server/src/aggregate.ts の SQL および
// android/.../domain/Aggregate.kt と完全に一致させること。**
// ズレると同じ週なのに違う数字が出る。

/**
 * その日が属する週の月曜日を返す。
 *
 * `%Y-%W` 相当でグルーピングしてはいけない（年をまたぐ週が分断される）。
 * @param {string} dateText YYYY-MM-DD
 * @returns {string} YYYY-MM-DD
 */
export function weekStart(dateText) {
  const d = new Date(dateText + "T00:00:00Z");
  // getUTCDay(): 0=日曜。月曜起点にするため (day + 6) % 7 日ぶん戻す
  const back = (d.getUTCDay() + 6) % 7;
  d.setUTCDate(d.getUTCDate() - back);
  return d.toISOString().slice(0, 10);
}

export function daysInMonth(key) {
  const [y, m] = key.split("-").map(Number);
  return new Date(Date.UTC(y, m, 0)).getUTCDate();
}

export function isLeapYear(y) {
  return (y % 4 === 0 && y % 100 !== 0) || y % 400 === 0;
}

export function daysInPeriod(key, period) {
  switch (period) {
    case "day": return 1;
    case "week": return 7;
    case "month": return daysInMonth(key);
    case "year": return isLeapYear(Number(key)) ? 366 : 365;
    default: return 1;
  }
}

/**
 * 日次の記録を期間ごとにまとめる。
 *
 * **平均の分母は記録がある日だけ。** 未計測の日を0歩として平均を下げない。
 * 0歩と記録された日は分母に含める。
 *
 * @param {Record<string, number>} stepsByDate 日付 -> 歩数（記録がある日だけ）
 * @param {"day"|"week"|"month"|"year"} period
 */
export function aggregate(stepsByDate, period) {
  const groups = new Map();
  for (const [dateText, steps] of Object.entries(stepsByDate)) {
    let key;
    switch (period) {
      case "day": key = dateText; break;
      case "week": key = weekStart(dateText); break;
      case "month": key = dateText.slice(0, 7); break;
      case "year": key = dateText.slice(0, 4); break;
    }
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(steps);
  }

  return [...groups.entries()]
    .map(([key, values]) => {
      const total = values.reduce((a, b) => a + b, 0);
      return {
        key,
        total,
        average: total / values.length,
        daysRecorded: values.length,
        daysInPeriod: daysInPeriod(key, period),
      };
    })
    .sort((a, b) => (a.key < b.key ? 1 : -1));
}

/** 目標に対する達成率。リング表示のため 1 で頭打ちにする。 */
export function achievedRatio(steps, goal) {
  if (goal <= 0) return 0;
  return Math.min(1, Math.max(0, steps / goal));
}

export function isAchieved(steps, goal) {
  return steps >= goal;
}

/**
 * 連続達成日数。
 *
 * **未計測の日は切らない**（端末を持たなかっただけかもしれない）。
 * **0歩と記録された日は切る**（実際に歩かなかったと確定している）。
 * **当日が未達でも切らない**（まだ挽回できる）。
 */
export function currentStreak(stepsByDate, todayText, goal) {
  let streak = 0;
  let date = todayText;

  const todaySteps = stepsByDate[todayText];
  if (todaySteps === undefined || !isAchieved(todaySteps, goal)) {
    date = shiftDate(todayText, -1);
  }

  const oldest = Object.keys(stepsByDate).sort()[0];
  if (oldest === undefined) return 0;

  while (date >= oldest) {
    const steps = stepsByDate[date];
    if (steps !== undefined) {
      if (isAchieved(steps, goal)) streak++;
      else return streak;      // 記録があって未達 → ここで途切れる
    }
    date = shiftDate(date, -1);
  }
  return streak;
}

/** これまでの最長の連続達成日数。 */
export function longestStreak(stepsByDate, goal) {
  const achieved = Object.keys(stepsByDate)
    .filter((d) => isAchieved(stepsByDate[d], goal))
    .sort();
  if (achieved.length === 0) return 0;

  let best = 1;
  let run = 1;
  for (let i = 1; i < achieved.length; i++) {
    // 間の日がすべて未計測なら連続とみなす
    let gapAllUnmeasured = true;
    for (let d = shiftDate(achieved[i - 1], 1); d < achieved[i]; d = shiftDate(d, 1)) {
      if (stepsByDate[d] !== undefined) { gapAllUnmeasured = false; break; }
    }
    run = gapAllUnmeasured ? run + 1 : 1;
    best = Math.max(best, run);
  }
  return best;
}

export function shiftDate(dateText, days) {
  const d = new Date(dateText + "T00:00:00Z");
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

/**
 * 直前の期間との平均の差。
 * **期間の途中では同じ経過日数までで比べる**（月初の不公平を避ける）。
 */
export function compareAverages(current, previous, elapsedDays) {
  const cur = Object.values(current);
  if (cur.length === 0) return null;

  const prev = Object.keys(previous).sort().slice(0, elapsedDays).map((k) => previous[k]);
  if (prev.length === 0) return null;

  const avg = (xs) => xs.reduce((a, b) => a + b, 0) / xs.length;
  return avg(cur) - avg(prev);
}
